### ISSUE TITLE
Feature Request: TOE Auto-Population

### ISSUE THREAD

**Issue by trent1098s** [MegaMek/mekhq#348]
To the TOE right click context menu add;

Change "Add Force..." so it extends to a sub menu.

"Single" (replicating what Add Force currently does, add a single child to the currently selected force)
----
IS Battlemech>
"Standard 'Mech Company" (adds a company with 3 lances; 1st Lance, 2nd Lance, 3rd Lance..)
"Standard 'Mech Battalion" (prompt for # of companies, then add 'Company A'...X)
"Standard 'Mech Regiment" (prompt for # of Battalions, then calls the above command X times)

IS Aerospace >
"Standard Squadron" (adds 3 lances)
"Standard Wing" (adds 3 squadrons of 3 lances)
"Standard Reinforced Wing"

IS Vehicle >
"Standard Company" (adds 3 platoons, 1st platoon, 2nd platoon, etc)
"Standard Squadron/Battalion" (prompt for # then add company A, Company B, etc)
"Standard Regiment" (prompt for # of squadrons then add 1st Squadron, 2nd Squadron, etc)

IS Infantry > 
As above...

IS Armor >
As above..

Clan Battlemech / Clan Aerospace / etc as above...

Comstar/WOB etc as above...

Bonus points if you can make it faction-specific based on what faction you have selected in the campaign options. (Per various field manuals)

Extra bonus points if you pre-create the correct layered icons. :)

**Comment by neoancient** [comment]
I think there are way too many possibilities to make this viable.

**Comment by trent1098s** [comment]
How so? Unit structure is a well defined concept for each faction. The layered unit icon system is complete enough to pull off doing icons. And the RAT tables exist to populate the unit by year.

I mean, everything is there that needs to be there to do something like this;

http://i.imgur.com/mLawq6m.png

or this..

http://i.imgur.com/DqKDLG8.png

Both of those groups were made for our gaming group (along with another campaign file for the player's unit). The setting is the rebellion at Zion in the mid to late 3050's, their merc unit drew a contract to assist FWL forces to help stabilize the province. (One of six possible contracts they had available).

Before we start the campaign (RPG + tactical) I created files for the factions in question. Canonically the 3rd Free Worlds Guard is involved (I'm splitting out a section of this in to "loyalists" and "rebels" since there was blue-on-blue fighting during the rebellion).  The 3rd Guard comprised a Battlemech regiment, an air wing, a heavy vehicle regiment, and an infantry regiment (all included in that second screenshot; total of 478 units including combat units and support).

The other unit involved was Carson's renegades. That unit was detailed in the Merc supplement.

Both use standard Inner Sphere regimental design, so share a common force structure. That's all I'm asking to create since it takes literally hours to build that TOE by hand. There's about 8 hours wrapped up in the FWL TOE screen alone, with more time populating out the various equipment. (For that I used the RATs in Megamek and exported the MUL file, then added units to the TOE from there). 

This very well could be something that could easily be scripted. Force structure is a known thing - regiments, battalions, companies, lances. Air wings, squadrons, flights. Leave it to the player to add in support elements as they see fit (artillery, HQ elements, etc).

Same holds true for comstar Level I, II, etc, and Clan stars, novas, galaxies.

It'd be a huge time saver for building up larger blocks of units for a campaign. Instead of a 12 hour process to make a regiment sized unit, it could all be done in a few seconds with the click of a button. 

(Since this campaign is a gritty / bloody one those various elements will be battered over time, and since there's a lot of RPG elements to this campaign I needed to know who had what, and where, so the players recon / specops team can go about their business).

Anyway, just a couple of examples of what I'm talking about. Not too hard to say "insert regiment" and have it prompt for # of battalions, # of companies per battalion, then create a TOE w/ 1 regiment, with X battalions, Y companies each, 3 lances per company.

**Comment by neoancient** [comment]
Are you asking for a blank TO&E that you fill in, or for the entire initial force to be generated? My response was based on the understanding that you wanted a blank TO&E, and the reason I said there were too many possibilities to make it viable is that while there may be a "most typical" structure for IS forces, there is considerable variation. Take, for instance, the FWL vehicle regiments, which have 4-6 units per lance, and 2-3 lances/company and 2-3 companies/"squadron". Or the FedCom/Lyran variation between reinforced and non-reinforced formations, which usually involve the addition of a command unit rather than a fourth sub-unit at the same eschelon as the others. Or the post-Clan Invasion development of Capellan augmented units, where lances can be mixes of 'Mechs and vehicles, or 'Mechs and BA, or BA and vehicles, and as time goes on they build up reinforced companies, battalions, and regiments. There is shift post-3050 across factions toward larger formations of BA in the infantry regiments. The LCAF shifted their organization to RCTs during the formation of the Federated Commonwealth. The DC Sword of Light uses four battalions per regiment instead of three. The FedSuns Ceti Hussars use combined arms formations at the battalion and company levels, as do the Outworlds Alliance in all their ground forces, and there are similar units in the LAAF and St. Ives Compact. None of this holds a candle to the amount of variation in the Clans. With the exception of Blood Spirit, there is no single cluster organization in a Clan, with some Clans varying more than others; second-line forces tend to have even more variation than front-line, and Fire Mandrill has a different range of possibilities depending on which Kindraa is under discussion.

Random generation is a different question. It takes the amount of variation into account, and it's something I've been working on on-and-off for the past three years. I still need to incorporate the formation building rules from Campaign Operations, and I'm not sure when I'll get back to working on it. It can handle any faction in any era from around the beginning of the Age of War to the Dark Age (though early in the Age of War there aren't enough primitive tanks to give very satisfactory results).

**Comment by trent1098s** [comment]
Yes just blank TO&E. For now populating the units is (comparably) not as time consuming of a task, since the RATs in Megamek can be used to generate a suitable force composition and size - and that's going to vary wildly from unit to unit. (E.g. Carson's Renegades, above, leans towards heavy mechs, while FWL was a medium mech regiment, and the fluff text indicates that heavies are very rare; spelling out certain regiments that have heavy assets assumes that the ones not indicated would not have them.)

By prompting for # of battalions / regiment, # of companies / battalion, # of lances per company, you could accommodate most inner sphere unit. Maybe a checkbox for "regimental HQ attachment" or "battalion artillery assets" or something, if you wanted to get more fancy.  

The player would still need to do some "fixing up" along the way, of course, since there's a wide variety, but at least the TOE would be generated (preferably with the force icons pre-made based on faction/type of formation). 

From that point at least the TOE is (mostly) done with a minimum of effort. The player would need to add in, say, a command flight of 2, for a FWL wing (manually), or "Add Company" to add an additional company to a regiment for an attached recon asset to the regiment, which falls outside of the normal battalion structure.

Clans would be done the same fashion, with options for Trinary, Supernova, etc in addition to the normal Novas.

Once the TOE is built the player can populate it using more conventional methods, as I did using MegaMek to build up my forces, then importing the MUL and assigning units. That doesn't take nearly as long. 

It's all about # of mouse clicks here. To build up a "standard IS regiment" is 13+ TOE entries per Battalion.  That's 3 clicks each. If you do a layered icon for each, that's an additional 15+ mouse clicks, per icon. (Varies depending on what you select, a basic one like this took 15 clicks)

![image](https://cloud.githubusercontent.com/assets/26397173/24074608/f1932fbe-0bd9-11e7-8d83-78354364ddcc.png)


By the time you finish a regiment, before you even GET to the point of adding units, you are hundreds of clicks and a couple hours in :)

Of course it'd be really wicked to have a RAT generator available right off the same menu; "GM Mode > Add Forces", to bring up the Megamek Add Forces.. dialog.. but this is a "one request per topic" thing so .. :)

**Comment by trent1098s** [comment]
Also to be clear this is mainly to support detailed player campaigns and not AtB type stuff. The forces I generated above would be used to deploy units for battles against humans, with the GM playing OpFor, using the MekHQ for the purpose of tracking and running the tabletop campaign.

Thinking longer term, however, if the foundation were laid for this type of larger unit construct, it would _potentially_ allow AtB to become something even greater if OpFor units on a mission were a "thing" that was tracked. The enemy would have finite resources, and beating the heck out of them would have noticeable effects on the arc of the AtB campaign. I remember keeping track of a Relief mission against Clan Jade Falcon in one of my AtB runs, and by the time it was all said and done my merc unit had faced (over several months of game time) multiple _galaxies_ of units in combat. 

Could add a tremendous amount of depth to contract negotiations to know if you're about to go up against a simple Merc company, or multiple Clan Galaxies. :)

(Again just long term dreaming there)

But it's all stepping stones, along the way. Each piece builds on the next. First step is TOE. :)
